src/safeposix/syscalls/fs_calls.rs

```diff
@@ -878,34 +1024,45 @@ impl Cage {
                     }
                     Inode::Socket(ref mut f) => {
                         f.linkcount -= 1;
+                        // Sockets only exist as long as the cages using them are running.
+                        // After these cages are closed, no changes to sockets' inodes
+                        // need to be persisted, thus using log is unnecessary and is set to "false"
                         (f.refcount, f.linkcount, false, false)
                     }
-                    Inode::Dir(_) => {
-                        return syscall_error(Errno::EISDIR, "unlink", "cannot unlink directory");
-                    }
                 }; //count current number of links and references
 
                 drop(inodeobj);
 
+                // Once the link count for the file has been decremented, we need to remove the
+                // reference of file from the parent directory. If the removal is successful,
+                // 0 is returned, otherwise an error with value!=0 is returned by the function.
                 let removal_result = Self::remove_from_parent_dir(parentinodenum, &truepath);
                 if removal_result != 0 {
                     return removal_result;
                 }
 
-                if curlinkcount == 0 {
-                    if currefcount == 0 {
-                        //actually remove file and the handle to it
-                        FS_METADATA.inodetable.remove(&inodenum);
-                        if has_fobj {
-                            let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
-                            interface::removefile(sysfilename).unwrap();
-                        }
-                    } //we don't need a separate unlinked flag, we can just
-                      // check that refcount is 0
+                // When the file's link count becomes 0 and no process has the file open
+                // (reference count = 0), the space occupied by the file will be
+                // freed and the file is no longer accessible.
+                if curlinkcount == 0 && currefcount == 0 {
```

Comment:
What if only one or the other is true?  When might that occur?

Reply:
There are two possible scenarios when link count = 0. If the refcount = 0, we delete the file from the system, but for the other case, we don't do anything extra. 
Added the scenarios in the comments and updated the code for better understanding.